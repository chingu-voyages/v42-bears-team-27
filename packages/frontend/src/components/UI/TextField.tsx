/* eslint-disable react/jsx-props-no-spreading */
import type {
  HTMLAttributes,
  RefAttributes,
  ForwardRefExoticComponent,
} from 'react';
import { forwardRef } from 'react';
import type { ThemeUIStyleObject } from 'theme-ui';
import * as Label from '@radix-ui/react-label';

export interface TextFieldProps
  extends HTMLAttributes<HTMLInputElement | HTMLTextAreaElement> {
  /**
   * Label for the component
   */
  label: string;
  /**
   * Type of the input element
   * @default 'text'
   */
  type?: 'text' | 'email' | 'password';
  /**
   * Enable multiline for the component
   * @default false
   */
  multiline?: boolean;
  sx?: ThemeUIStyleObject;
  value?: string;
}

export const TextField: ForwardRefExoticComponent<
TextFieldProps & RefAttributes<any>
> = forwardRef(
  ({ id, label, type = 'text', multiline = false, sx, ...rest }, ref) => {
    const styles: ThemeUIStyleObject = {
      variant: 'text.label',
      width: 'inherit',
      height: 'inherit',
      py: 3,
      pl: 26,
      pr: 4,
      fontFamily: 'heading',
      fontSize: 14,
      color: 'primary',
      border: '2px solid',
      borderColor: 'gray',
      '&:placeholder-shown': {
        color: 'gray',
      },
    };

    return (
      <div sx={{ display: 'flex', flexDirection: 'column', rowGap: 3, ...sx }}>
        <Label.Root
          htmlFor={id}
          sx={{ variant: 'text.label', fontFamily: 'heading', fontSize: 18 }}
        >
          {label}
        </Label.Root>
        <div
          sx={{
            width: '100%',
            height: multiline ? 174 : 50,
            bg: 'background',
          }}
        >
          {multiline ? (
            <textarea ref={ref} id={id} sx={styles} {...rest} />
          ) : (
            <input ref={ref} id="firstName" type={type} sx={styles} {...rest} />
          )}
        </div>
      </div>
    );
  },
);

if (process.env.NODE_ENV !== 'production') {
  TextField.displayName = 'TextField';
}
