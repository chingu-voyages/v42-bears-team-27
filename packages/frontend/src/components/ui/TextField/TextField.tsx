import type {
  HTMLAttributes,
  RefAttributes,
  ForwardRefExoticComponent,
} from 'react';
import { forwardRef } from 'react';
import type { ThemeUIStyleObject } from 'theme-ui';
import * as LabelPrimitive from '@radix-ui/react-label';

export interface TextFieldProps
  extends HTMLAttributes<HTMLInputElement | HTMLTextAreaElement> {
  /**
   * Label for the component
   */
  label: string;
  /**
   * Type of the input element
   * @default 'text'
   */
  type?: 'text' | 'email' | 'password';
  /**
   * Enable multiline for the component
   * @default false
   */
  multiline?: boolean;
  sx?: ThemeUIStyleObject;
  value?: string;
  required?: boolean;
}

export const TextField: ForwardRefExoticComponent<
TextFieldProps & RefAttributes<any>
> = forwardRef(
  (
    {
      key,
      id,
      label,
      type = 'text',
      multiline = false,
      required = false,
      sx,
      ...rest
    },
    ref,
  ) => {
    const styles: ThemeUIStyleObject = {
      variant: 'text.label',
      width: 'inherit',
      height: 'inherit',
      py: 3,
      pl: 26,
      pr: 4,
      fontFamily: 'heading',
      fontSize: 14,
      color: 'primary',
      border: '2px solid',
      borderColor: 'gray',
      '&:placeholder-shown': {
        color: 'gray',
      },
    };

    return (
      <div
        key={key}
        sx={{ display: 'flex', flexDirection: 'column', rowGap: 3, ...sx }}
      >
        <LabelPrimitive.Root
          htmlFor={id}
          sx={{ variant: 'text.label', fontFamily: 'heading', fontSize: 18 }}
        >
          {label}
        </LabelPrimitive.Root>
        <div
          sx={{
            width: '100%',
            height: multiline ? 174 : 50,
            bg: 'background',
          }}
        >
          {multiline ? (
            <textarea
              ref={ref}
              id={id}
              required={required}
              sx={styles}
              {...rest}
            />
          ) : (
            <input
              ref={ref}
              id={id}
              type={type}
              required={required}
              sx={styles}
              {...rest}
            />
          )}
        </div>
      </div>
    );
  },
);

if (process.env.NODE_ENV !== 'production') {
  TextField.displayName = 'TextField';
}
