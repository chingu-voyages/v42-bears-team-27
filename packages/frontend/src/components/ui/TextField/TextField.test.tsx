import { render, screen } from 'test-utils';

import { TextField } from './TextField';

describe('<TextField />', () => {
  describe('Snapshots', () => {
    it('should match snapshot', () => {
      const { container } = render(<TextField label="Foo" />);
      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with multiline prop', () => {
      const { container } = render(<TextField label="Foo" multiline />);
      expect(container).toMatchSnapshot();
    });
  });

  describe('label prop', () => {
    it('should display label as TextField label text', () => {
      render(<TextField id="foo" label="Foo" />);
      expect(screen.getByLabelText('Foo'));
    });
  });
});
