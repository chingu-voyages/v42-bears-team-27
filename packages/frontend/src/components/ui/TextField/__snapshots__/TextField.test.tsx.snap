// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> Snapshots should match snapshot 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  row-gap: 16px;
}

.emotion-1 {
  font-family: "Roboto",sans-serif;
  font-size: 18px;
  font-weight: 400;
  line-height: 1.5;
}

.emotion-2 {
  width: 100%;
  height: 50px;
  background-color: var(--theme-ui-colors-background);
}

.emotion-3 {
  font-family: "Roboto",sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1.5;
  width: inherit;
  height: inherit;
  padding-top: 16px;
  padding-bottom: 16px;
  padding-left: 26px;
  padding-right: 32px;
  color: var(--theme-ui-colors-primary);
  border: 2px solid;
  border-color: var(--theme-ui-colors-gray);
}

.emotion-3:placeholder-shown {
  color: var(--theme-ui-colors-gray);
}

<div>
  <div
    class="emotion-0"
  >
    <label
      class="emotion-1"
    >
      Foo
    </label>
    <div
      class="emotion-2"
    >
      <input
        class="emotion-3"
        type="text"
      />
    </div>
  </div>
</div>
`;

exports[`<TextField /> Snapshots should match snapshot with multiline prop 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  row-gap: 16px;
}

.emotion-1 {
  font-family: "Roboto",sans-serif;
  font-size: 18px;
  font-weight: 400;
  line-height: 1.5;
}

.emotion-2 {
  width: 100%;
  height: 174px;
  background-color: var(--theme-ui-colors-background);
}

.emotion-3 {
  font-family: "Roboto",sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1.5;
  width: inherit;
  height: inherit;
  padding-top: 16px;
  padding-bottom: 16px;
  padding-left: 26px;
  padding-right: 32px;
  color: var(--theme-ui-colors-primary);
  border: 2px solid;
  border-color: var(--theme-ui-colors-gray);
}

.emotion-3:placeholder-shown {
  color: var(--theme-ui-colors-gray);
}

<div>
  <div
    class="emotion-0"
  >
    <label
      class="emotion-1"
    >
      Foo
    </label>
    <div
      class="emotion-2"
    >
      <textarea
        class="emotion-3"
      />
    </div>
  </div>
</div>
`;
